pkg/dcmtime/precision.go

```diff
@@ -0,0 +1,74 @@
+package dcmtime
+
+// PrecisionLevel tells encoding methods how many segments of the value to render
+// to the DICOM format string, as the spec allows eliding segments to communicate the
+// precision.
+//
+// For instance: when rendering a DA (date) value, using a truncation of Precision.Year
+// would render the entire date as 'YYYY'.
+//
+// Using Precision.Month would render 'YYYYMM'.
+//
+// Using Precision.Day or Precision.Full would render a full 'YYYYMMDD'.
+type PrecisionLevel int
+
+// String returns the name of the PrecisionLevel for debugging.
+func (level PrecisionLevel) String() string {
+	switch level {
+	case Precision.Full:
+		return "FULL"
+	case Precision.Year:
+		return "YEAR"
+	case Precision.Month:
+		return "MONTH"
+	case Precision.Day:
+		return "DAY"
+	case Precision.Hours:
+		return "HOURS"
+	case Precision.Minutes:
+		return "MINUTES"
+	case Precision.Seconds:
+		return "SECONDS"
+	case Precision.MS1:
+		return "MS1"
+	case Precision.MS2:
+		return "MS2"
+	case Precision.MS3:
+		return "MS3"
+	case Precision.MS4:
+		return "MS4"
+	case Precision.MS5:
+		return "MS5"
+	default:
+		return "!{PRECISION}"
+	}
+}
+
+// Precision is a namespace for PrecisionLevel values.
+var Precision = struct {
```

Comment:
I'm not sure how I feel about this, it's not a common pattern I've seen elsewhere. Usually for enums I see something like 

```go
const (
    PrecisionYear PrecisionLevel = iota
    PrecisionMonth
    ....
)
```
I think the small benefit with the above is that they are constant, and cannot be accidentally changed. during the runtime of the program (e.g. by doing something like `dcmtime.Precision.Year = X` (not that anyone should be doing something like this)).

A bit annoying that structs can be const in Go (https://golang.org/ref/spec#Constants)

However I do kind of like the way this reads, so a little on the fence here. It's probably fine as is, but it might be better to stick with the convention seen for enums elsewhere. Thoughts?

Reply:
I think you're right that we should follow conventions here. Every now and then I think I get frustrated with Go's lack of encapsulated enums and try out something like this, but I should not be fighting the language.

HARD agree on wishing that structs could be const. Maybe if this proposal is ever accepted and implemented: https://github.com/golang/go/issues/32245